### ISSUE TITLE
keras save model error

### ISSUE THREAD

**Issue by sunshinenum** [tensorflow/recommenders-addons#172]
**System information**
- OS Platform and Distribution
    Fedora 34
- TensorFlow version and how it was installed (source or binary):
   2.5.1 CPU, build from source
- TensorFlow-Recommenders-Addons version and how it was installed (source or binary):
  0.3.0
- Python version:
  3.8.3
- Is GPU used? (yes/no):
  no

**Describe the bug**

  save model error with keras api.

**Code to reproduce the issue**

  **embedding define (in a self-define layer):**
```
          self.user_embeddings = tfra.dynamic_embedding.get_variable(
            name="user_dynamic_embeddings",
            dim=8,
            devices="CPU:0",
            initializer=tf.keras.initializers.Zeros())
```
  **embedding call:**
```
        user_id_hashed = tf.strings.to_hash_bucket_fast(features["f003.uid"], 500000000)
        user_id_val, user_id_idx = tf.unique(tf.concat(user_id_hashed, axis=0))
        user_id_weights, user_id_trainable_wrapper = tfra.dynamic_embedding.embedding_lookup(
            params=self.user_embeddings,
            ids=user_id_val,
            name="user-id-weights",
            return_trainable=True)
        user_id_weights = tf.gather(user_id_weights, user_id_idx)
```
  **model save**
```
model.save_weights(os.path.join(self.model_dir, 'weights')) #  ok
```
```
model.save(saving_dir, signatures=signatures, include_optimizer=False, options=tf.saved_model.SaveOptions(namespace_whitelist=['TFRA']) # error
```

**error*

AssertionError: Called a function referencing variables which have been deleted. This likely means that function-local variables were created and not referenced elsewhere in the program. This is generally a mistake; consider storing variables in an object attribute on first call.

**Other info / logs**

2021-10-17 12:35:26,129 : model weights saved !
WARNING:tensorflow:Skipping full serialization of Keras layer <tensorflow.python.keras.layers.embeddings.Embedding object at 0x7f5b00aae250>, because it is not built.
2021-10-17 12:35:32,714 : Skipping full serialization of Keras layer <tensorflow.python.keras.layers.embeddings.Embedding object at 0x7f5b00aae250>, because it is not built.
Traceback (most recent call last):
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/carlos/Downloads/models82/wdl/train_base.py", line 151, in <module>
    fire.Fire(main)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/fire/core.py", line 466, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/fire/core.py", line 681, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "/home/carlos/Downloads/models82/wdl/train_base.py", line 146, in main
    base_trainer.work()
  File "/home/carlos/Downloads/models82/wdl/train_base.py", line 71, in work
    self.train_by_file_names(file_names1)
  File "/home/carlos/Downloads/models82/wdl/train_base.py", line 97, in train_by_file_names
    self.train(file_names[-1:])
  File "/home/carlos/Downloads/models82/wdl/utils/rank_models.py", line 111, in train
    model.save(saving_dir, signatures=signatures, include_optimizer=False)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/keras/engine/training.py", line 2106, in save
    save.save_model(self, filepath, overwrite, include_optimizer, save_format,
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/keras/saving/save.py", line 150, in save_model
    saved_model_save.save(model, filepath, overwrite, include_optimizer,
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/keras/saving/saved_model/save.py", line 89, in save
    saved_nodes, node_paths = save_lib.save_and_return_nodes(
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/saved_model/save.py", line 1103, in save_and_return_nodes
    _build_meta_graph(obj, signatures, options, meta_graph_def,
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/saved_model/save.py", line 1290, in _build_meta_graph
    return _build_meta_graph_impl(obj, signatures, options, meta_graph_def,
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/saved_model/save.py", line 1212, in _build_meta_graph_impl
    signature_serialization.validate_saveable_view(checkpoint_graph_view)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/saved_model/signature_serialization.py", line 301, in validate_saveable_view
    for name, dep in saveable_view.list_dependencies(
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/saved_model/save.py", line 120, in list_dependencies
    extra_dependencies = self.list_extra_dependencies(obj)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/saved_model/save.py", line 148, in list_extra_dependencies
    return obj._list_extra_dependencies_for_serialization(  # pylint: disable=protected-access
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/keras/engine/base_layer.py", line 3012, in _list_extra_dependencies_for_serialization
    return (self._trackable_saved_model_saver
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/keras/saving/saved_model/base_serialization.py", line 76, in list_extra_dependencies_for_serialization
    return self.objects_to_serialize(serialization_cache)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/keras/saving/saved_model/layer_serialization.py", line 69, in objects_to_serialize
    return (self._get_serialized_attributes(
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/keras/saving/saved_model/layer_serialization.py", line 89, in _get_serialized_attributes
    object_dict, function_dict = self._get_serialized_attributes_internal(
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/keras/saving/saved_model/model_serialization.py", line 53, in _get_serialized_attributes_internal
    super(ModelSavedModelSaver, self)._get_serialized_attributes_internal(
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/keras/saving/saved_model/layer_serialization.py", line 99, in _get_serialized_attributes_internal
    functions = save_impl.wrap_layer_functions(self.obj, serialization_cache)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/keras/saving/saved_model/save_impl.py", line 204, in wrap_layer_functions
    fn.get_concrete_function()
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/contextlib.py", line 120, in __exit__
    next(self.gen)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/keras/saving/saved_model/save_impl.py", line 367, in tracing_scope
    fn.get_concrete_function(*args, **kwargs)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/eager/def_function.py", line 1367, in get_concrete_function
    concrete = self._get_concrete_function_garbage_collected(*args, **kwargs)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/eager/def_function.py", line 1284, in _get_concrete_function_garbage_collected
    concrete = self._stateful_fn._get_concrete_function_garbage_collected(  # pylint: disable=protected-access
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/eager/function.py", line 3100, in _get_concrete_function_garbage_collected
    graph_function, _ = self._maybe_define_function(args, kwargs)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/eager/function.py", line 3444, in _maybe_define_function
    graph_function = self._create_graph_function(args, kwargs)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/eager/function.py", line 3279, in _create_graph_function
    func_graph_module.func_graph_from_py_func(
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/framework/func_graph.py", line 999, in func_graph_from_py_func
    func_outputs = python_func(*func_args, **func_kwargs)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/eager/def_function.py", line 672, in wrapped_fn
    out = weak_wrapped_fn().__wrapped__(*args, **kwds)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/keras/saving/saved_model/save_impl.py", line 599, in wrapper
    ret = method(*args, **kwargs)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/keras/saving/saved_model/utils.py", line 165, in wrap_with_training_arg
    return control_flow_util.smart_cond(
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/keras/utils/control_flow_util.py", line 109, in smart_cond
    return smart_module.smart_cond(
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/framework/smart_cond.py", line 56, in smart_cond
    return false_fn()
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/keras/saving/saved_model/utils.py", line 167, in <lambda>
    lambda: replace_training_and_call(False))
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/keras/saving/saved_model/utils.py", line 163, in replace_training_and_call
    return wrapped_call(*args, **kwargs)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/keras/saving/saved_model/save_impl.py", line 681, in call
    return call_and_return_conditional_losses(inputs, *args, **kwargs)[0]
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/keras/saving/saved_model/save_impl.py", line 639, in __call__
    return self.wrapped_call(*args, **kwargs)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/eager/def_function.py", line 889, in __call__
    result = self._call(*args, **kwds)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/eager/def_function.py", line 924, in _call
    results = self._stateful_fn(*args, **kwds)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/eager/function.py", line 3023, in __call__
    return graph_function._call_flat(
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/eager/function.py", line 1915, in _call_flat
    for v in self._func_graph.variables:
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/framework/func_graph.py", line 495, in variables
    return tuple(deref(v) for v in self._weak_variables)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/framework/func_graph.py", line 495, in <genexpr>
    return tuple(deref(v) for v in self._weak_variables)
  File "/home/chenliguo/tools/anaconda3/lib/python3.8/site-packages/tensorflow/python/framework/func_graph.py", line 487, in deref
    raise AssertionError(
AssertionError: Called a function referencing variables which have been deleted. This likely means that function-local variables were created and not referenced elsewhere in the program. This is generally a mistake; consider storing variables in an object attribute on first call.

**Comment by npbool** [comment]
I don't know much about tfra internals but from the code below I guess it's because TrainableWrapper is created again in exporting so reference to the original variable is lost.
![image](https://user-images.githubusercontent.com/1175364/138022191-f40e797d-e49c-4d9b-9b01-2858e983bea9.png)

**Comment by Lifann** [comment]
Sorry for the miss-spelling.

1. In eager mode, the TensorFlow variable including TrainableWrapper will behold in the life cycle as its Python object. If the TrainableWrapper was created in some function, it will be deleted out of the function scope. So when calling model.save, the TrainableWrapper will not be found, thus raise an `variables were created and not referenced elsewhere in the program` error. Try hold the TrainableWrapper in execution like [this](https://github.com/tensorflow/recommenders-addons/blob/master/demo/dynamic_embedding/amazon-us-reviews-digital-video-games/video_game_model.py#L65-L66)

2. Switch self.user_embedding.lookup(user_id_val) before saving the model. The TrainableWrapper is not need in inference. Need to avoid saving the TrainableWrapper to a saved_model. Please check [this](https://github.com/tensorflow/recommenders-addons/blob/811191ed7be779b6874af47c26d0b316a3963ac2/demo/dynamic_embedding/amazon-us-reviews-digital-video-games/video_game_model.py#L125-L150)

**Comment by npbool** [comment]
> Sorry for the miss-spelling.
> 
>     1. In eager mode, the TensorFlow variable including TrainableWrapper will behold in the life cycle as its Python object. If the TrainableWrapper was created in some function, it will be deleted out of the function scope. So when calling `model.save`, the TrainableWrapper will not be found, thus raise a `variables were created and not referenced elsewhere in the program` error. Try to hold the TrainableWrapper in execution like [this](https://github.com/tensorflow/recommenders-addons/blob/master/demo/dynamic_embedding/amazon-us-reviews-digital-video-games/video_game_model.py#L65-L66)
> 
>     2. Switch self.user_embedding.lookup(user_id_val) before saving the model. The TrainableWrapper is not need in inference. Need to avoid saving the TrainableWrapper to a saved_model. Please check [this](https://github.com/tensorflow/recommenders-addons/blob/811191ed7be779b6874af47c26d0b316a3963ac2/demo/dynamic_embedding/amazon-us-reviews-digital-video-games/video_game_model.py#L125-L150)

this works for me.
But to clone a separate `VideoGameDnnInference` seems awkward. I tried to add a `@tf.function` in `VideoGameDnn` for inference and saving but it fails like

`TypeError: An op outside of the function building code is being passed
    a "Graph" tensor. It is possible to have Graph tensors
    leak out of the function building context by including a
    tf.init_scope in your function building code.
    For example, the following function will fail:
      @tf.function
      def has_init_scope():
        my_constant = tf.constant(1.)
        with tf.init_scope():
          added = my_constant * 2
`

**Comment by Lifann** [comment]
> > Sorry for the miss-spelling.
> > ```
> > 1. In eager mode, the TensorFlow variable including TrainableWrapper will behold in the life cycle as its Python object. If the TrainableWrapper was created in some function, it will be deleted out of the function scope. So when calling `model.save`, the TrainableWrapper will not be found, thus raise a `variables were created and not referenced elsewhere in the program` error. Try to hold the TrainableWrapper in execution like [this](https://github.com/tensorflow/recommenders-addons/blob/master/demo/dynamic_embedding/amazon-us-reviews-digital-video-games/video_game_model.py#L65-L66)
> > 
> > 2. Switch self.user_embedding.lookup(user_id_val) before saving the model. The TrainableWrapper is not need in inference. Need to avoid saving the TrainableWrapper to a saved_model. Please check [this](https://github.com/tensorflow/recommenders-addons/blob/811191ed7be779b6874af47c26d0b316a3963ac2/demo/dynamic_embedding/amazon-us-reviews-digital-video-games/video_game_model.py#L125-L150)
> > ```
> 
> this works for me. But to clone a separate `VideoGameDnnInference` seems awkward. I tried to add a `@tf.function` in `VideoGameDnn` for inference and saving but it fails like
> 
> `TypeError: An op outside of the function building code is being passed a "Graph" tensor. It is possible to have Graph tensors leak out of the function building context by including a tf.init_scope in your function building code. For example, the following function will fail: @tf.function def has_init_scope(): my_constant = tf.constant(1.) with tf.init_scope(): added = my_constant * 2 `

`tf.function` is not supported yet. Here use the uncommon usage is because we have to bypass the issue(#167 ).

**Comment by Lifann** [comment]
Hi, @sunshinenum , tf.function and keras is currently supported in latest release (0.4.0)

**Comment by sunshinenum** [comment]
fixed in 0.4.0
